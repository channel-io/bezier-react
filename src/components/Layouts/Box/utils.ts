/* External dependencies */
import { CSSObject } from 'styled-components'
import { isNil, isNumber, isString, omit } from 'lodash-es'

/* Internal dependencies */
import { Foundation } from 'Foundation'
import { omittedPropKeys } from './const'

/**
 * @link https://styled-components.com/docs/advanced#style-objects
 * @description Style Objects로 prop을 전달받으면서 불필요한 prop도 함께 섞여 들어옵니다. 이를 삭제합니다.
 */
export function getSanitizedCSSObject(props: CSSObject) {
  return omit(props, omittedPropKeys)
}

export function convertThemeProps(foundation?: Foundation, ...values: unknown[]) {
  for (const index in values) {
    if (Object.prototype.hasOwnProperty.call(values, index)) {
      const eachValue = values[index]
      if (!isNil(eachValue)) {
        if (isString(eachValue)) {
          return foundation?.theme?.[eachValue] ?? eachValue
        }
      }
    }
  }
  return undefined
}

export function convertSpacingProps(foundation?: Foundation, ...values: unknown[]) {
  for (const index in values) {
    if (Object.prototype.hasOwnProperty.call(values, index)) {
      const eachValue = values[index]
      if (!isNil(eachValue)) {
        if (isNumber(eachValue)) {
          return eachValue === 0 ? eachValue : `${eachValue}px`
        }
        if (isString(eachValue)) {
          return foundation?.spacing?.[eachValue] ?? eachValue
        }
      }
    }
  }
  return undefined
}
