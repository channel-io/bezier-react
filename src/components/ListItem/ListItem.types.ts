/* Internal dependencies */
import React from 'react'
import { SemanticNames } from '../../foundation/Colors/Theme'
import ActivableElement from '../../types/ActivatableElement'
import { ContentComponentProps } from '../../types/ComponentProps'
import OptionItem from '../../types/OptionItem'
import { IconName } from '../Icon'

export enum ListItemSize {
  S = 's',
  M = 'm',
  L = 'l',
  XL = 'xl',
}

export default interface ListItemProps extends ContentComponentProps, ActivableElement, OptionItem {
  iconClassName?: string
  contentClassName?: string
  size?: ListItemSize
  name?: string
  href?: string
  hide?: boolean
  nested?: boolean
  leftContent?: React.ReactNode
  leftIcon?: IconName
  leftIconColor?: SemanticNames
  disableIconActive?: boolean
  descriptionMaxLines?: number
  content?: React.ReactNode
  description?: React.ReactNode
  rightContent?: React.ReactNode
  onClick?: (e: React.MouseEvent, name?: string) => void
  onMouseDown?: (e: React.MouseEvent) => void
  onMouseEnter?: (e: React.MouseEvent) => void
  onMouseLeave?: (e: React.MouseEvent) => void
}
