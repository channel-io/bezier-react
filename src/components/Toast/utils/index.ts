import { css } from '../../../foundation'
import { Appearance, IconColor, Placement } from '../Toast.types'

function getIconColor(appearance: Appearance): string {
  switch (appearance) {
    case Appearance.Success:
      return IconColor.Success
    case Appearance.Warning:
      return IconColor.Warning
    case Appearance.Error:
      return IconColor.Error
    case Appearance.Info:
    default:
      return IconColor.Info
  }
}

function placements(placement: Placement) {
  switch (placement) {
    case Placement.TopCenter:
      return css`
          top: 0;
          left: 50%;
          transform: translateX(-50%);
        `
    case Placement.TopRight:
      return css`
          top: 0;
          right: 0;
        `
    case Placement.TopLeft:
      return css`
          top: 0;
          left: 0;
        `
    case Placement.BottomCenter:
      return css`
          bottom: 0;
          left: 50%;
          transform: translateX(-50%);
        `
    case Placement.BottomRight:
      return css`
          right: 0;
          bottom: 0;
        `
    case Placement.BottomLeft:
    default:
      return css`
        bottom: 0;
        left: 0;
      `
  }
}

const initPosition = (placement: Placement, isXPosition: boolean) => {
  switch (placement) {
    case Placement.TopLeft:
      return isXPosition ? '-120%' : '0'
    case Placement.TopCenter:
      return isXPosition ? '0' : '-120%'
    case Placement.TopRight:
      return isXPosition ? '120%' : '0'
    case Placement.BottomLeft:
      return isXPosition ? '-120%' : '0'
    case Placement.BottomCenter:
      return isXPosition ? '0' : '120%'
    case Placement.BottomRight:
      return isXPosition ? '120%' : '0'
    default:
      return '0'
  }
}

export {
  getIconColor,
  placements,
  initPosition,
}
