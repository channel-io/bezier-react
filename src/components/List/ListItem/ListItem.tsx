/* External dependencies */
import React, { Ref, forwardRef, useCallback, useMemo } from 'react'
import { get, noop, isNil } from 'lodash-es'

/* Internal dependencies */
import { mergeClassNames } from '../../../utils/stringUtils'
import ListItemProps from './ListItem.types'
import { Wrapper } from './ListItem.styled'

export const SIDEBAR_MENU_ITEM_COMPONENT_NAME = 'ListItem'
export const SIDEBAR_MENU_ITEM_TEST_ID = 'ch-design-system-sidebar-menu-item'

export function isListItem(element: any): element is React.ReactElement<ListItemProps> {
  return React.isValidElement(element) &&
    get(element, 'type.displayName') === SIDEBAR_MENU_ITEM_COMPONENT_NAME
}

function ListItemComponent({
  as,
  testId = SIDEBAR_MENU_ITEM_TEST_ID,
  content,
  name,
  href,
  hide,
  /* OptionItem Props */
  optionKey,
  /* Activable Element Props */
  active = false,
  activeClassName,
  /* HTMLAttribute Props */
  onClick = noop,
  className,
  ...othreProps
}: ListItemProps, forwardedRef: Ref<any>) {
  const clazzName = useMemo(() => (
    mergeClassNames(className, ((active && activeClassName) || undefined))
  ), [
    className,
    activeClassName,
    active,
  ])

  const handleClick = useCallback((e) => {
    if (!active) {
      onClick(e, name)
    }
  }, [
    active,
    name,
    onClick,
  ])

  if (hide) return null

  if (!isNil(href)) {
    return (
      <Wrapper
        ref={forwardedRef}
        as="a"
        className={clazzName}
        draggable={false}
        href={href}
        target="_blank"
        rel="noopener noreferer"
        onClick={handleClick}
        active={active}
        data-active={active}
        data-option-key={optionKey}
        data-testid={testId}
        {...othreProps}
      >
        { content }
      </Wrapper>
    )
  }

  return (
    <Wrapper
      as={as}
      className={clazzName}
      onClick={handleClick}
      active={active}
      data-active={active}
      data-option-key={optionKey}
      data-testid={testId}
      {...othreProps}
    >
      { content }
    </Wrapper>
  )
}

const ListItem = forwardRef(ListItemComponent)
ListItem.displayName = SIDEBAR_MENU_ITEM_COMPONENT_NAME

export default ListItem
