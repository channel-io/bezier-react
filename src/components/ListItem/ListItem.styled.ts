/* Internal dependencies */
import disabledOpacity from '../../constants/DisabledOpacity'
import {
  css,
  ellipsis,
  LineHeightAbsoluteNumber,
  styled,
  TransitionDuration,
} from '../../foundation'
import { SemanticNames } from '../../foundation/Colors/Theme'
import { Icon } from '../Icon'
import {
  ListItemColorVariant,
  ListItemSize,
} from './ListItem.types'
import { getStyleOfSize } from './utils'

interface StyledWrapperProps {
  size?: ListItemSize
  active?: boolean
  focused?: boolean
  disabled?: boolean
  colorVariant: ListItemColorVariant
}

const getColorFromColorVariantWithDefaultValue = (
  colorVariant: ListItemColorVariant,
  defaultValue: SemanticNames,
): SemanticNames => {
  switch (colorVariant) {
    case (ListItemColorVariant.Blue):
      return 'bgtxt-blue-normal'
    case (ListItemColorVariant.Red):
      return 'bgtxt-red-normal'
    case (ListItemColorVariant.Green):
      return 'bgtxt-green-normal'
    case (ListItemColorVariant.Cobalt):
      return 'bgtxt-cobalt-normal'
    default:
      return defaultValue
  }
}

const FocusedItemStyle = css<StyledWrapperProps>`
  background-color: ${({ foundation }) => foundation?.theme?.['bg-black-lighter']};
`

const FocusedIconStyle = css<StyledWrapperProps>`
  color: ${({ foundation, active, colorVariant }) => (
    foundation?.theme?.[getColorFromColorVariantWithDefaultValue(
      colorVariant,
      active ? 'bgtxt-blue-normal' : 'txt-black-dark',
    )]
  )};
`

const ActiveItemStyle = css<StyledWrapperProps>`
  color: ${({ foundation, colorVariant }) => (
    foundation?.theme?.[getColorFromColorVariantWithDefaultValue(colorVariant, 'bgtxt-blue-normal')]
  )};
  background-color: ${({ foundation }) => foundation?.theme?.['bgtxt-blue-lightest']};
`

export const RightContent = styled.div`
  margin-left: 8px;
`

export const TitleWrapper = styled.div`
  display: flex;
  grid-row: 1;
  grid-column: 2;
  align-items: center;

  & span {
    ${({ foundation }) => foundation?.transition?.getTransitionsCSS('color')};
  }
`

export const DescriptionWrapper = styled.div`
  display: flex;
  grid-row: 2;
  grid-column: 2;
  align-items: center;
  width: 100%;
  margin-top: 2px;
  color: ${({ foundation }) => (foundation?.theme['txt-black-darker'])
};
`

export const Title = styled.div`
  ${ellipsis()}
`

interface DescriptionProps {
  descriptionMaxLines?: number
}

export const Description = styled.div<DescriptionProps>`
  ${({ descriptionMaxLines }) => descriptionMaxLines && ellipsis(descriptionMaxLines, LineHeightAbsoluteNumber.Lh16)}
`

interface IconWrapperProps {
  colorVariant: ListItemColorVariant
  active?: boolean
}

export const StyledIcon = styled(Icon)<IconWrapperProps>`
  color: ${({ foundation, colorVariant, active }) => (
    foundation?.theme?.[getColorFromColorVariantWithDefaultValue(colorVariant, active ? 'bgtxt-blue-normal' : 'txt-black-dark')]
  )};
  
  ${({ foundation }) => foundation?.transition?.getTransitionsCSS('color', TransitionDuration.M)};
`

export const LeftContentWrapper = styled.div`
  display: flex;
  grid-row: 1;
  grid-column: 1;
  align-items: center;
  margin-right: 8px;
`

export const ContentWrapper = styled.div`
  display: grid;
  grid-template-columns: fit-content(100%) minmax(0, 1fr);
  width: 100%;
`

export const Wrapper = styled.div<StyledWrapperProps>`
  display: flex;
  align-items: center;
  ${({ size }) => getStyleOfSize(size)}

  color: ${({ foundation, colorVariant }) => (
    foundation?.theme?.[getColorFromColorVariantWithDefaultValue(colorVariant, 'txt-black-darkest')]
  )};
  
  text-decoration: none;
  cursor: ${({ disabled }) => (disabled ? 'default' : 'pointer')};
  opacity: ${({ disabled }) => (disabled ? disabledOpacity : 1)};

  ${({ foundation }) => foundation?.transition?.getTransitionsCSS(['background-color'])};

  ${({ disabled, active, focused }) => !disabled && !active && css`
    ${focused && css`
      ${FocusedItemStyle}

      ${StyledIcon} {
        ${FocusedIconStyle}
      }
    `}

    &:hover {
      ${FocusedItemStyle}
    }

    &:hover ${StyledIcon} {
      ${FocusedIconStyle}
    }
`};

  ${({ active }) => (active && ActiveItemStyle)}
`
