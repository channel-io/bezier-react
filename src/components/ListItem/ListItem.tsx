/* External dependencies */
import React, { Ref, forwardRef, useCallback, useMemo, Fragment } from 'react'
import { get, noop, isNil, isString } from 'lodash-es'
import { v4 as uuid } from 'uuid'

/* Internal dependencies */
import { mergeClassNames } from '../../utils/stringUtils'
import { Text } from '../Text'
import { Icon, IconSize } from '../Icon'
import { isIconName } from '../Icon/util'
import { Typography } from '../../foundation'
import ListItemProps, { ListItemSize } from './ListItem.types'
import {
  Wrapper,
  ContentWrapper,
  IconWrapper,
  IconMargin,
  TitleWrapper,
  Title,
  DescriptionWrapper,
  Description,
  RightContent,
} from './ListItem.styled'

export const LIST_ITEM_COMPONENT_NAME = 'ListItem'
export const LIST_ITEM_TEST_ID = 'ch-design-system-list-menu-item'

export function isListItem(element: any): element is React.ReactElement<ListItemProps> {
  return React.isValidElement(element) &&
    get(element, 'type.displayName') === LIST_ITEM_COMPONENT_NAME
}

function ListItemComponent({
  className,
  contentClassName,
  iconClassName,
  as,
  testId = LIST_ITEM_COMPONENT_NAME,
  size = ListItemSize.M,
  descriptionMaxLines,
  content,
  description,
  name,
  leftIcon,
  leftIconColor,
  disableIconActive = false,
  href,
  hide = false,
  rightContent = null,
  /* OptionItem Props */
  optionKey,
  /* Activable Element Props */
  active,
  activeClassName,
  /* HTMLAttribute Props */
  onClick = noop,
  ...othreProps
}: ListItemProps, forwardedRef: Ref<any>) {
  const clazzName = useMemo(() => (
    mergeClassNames(className, ((active && activeClassName) || undefined))
  ), [
    className,
    activeClassName,
    active,
  ])

  const handleClick = useCallback((e) => {
    if (!active) {
      onClick(e, name)
    }
  }, [
    active,
    name,
    onClick,
  ])

  const getNewLineComponenet = useCallback((desc: string) => (
    desc.split('\n').map((str, index) => {
      if (index === 0) {
        return (
          <Text key={uuid()} typo={Typography.Size14}>
            { str }
          </Text>
        )
      }

      return (
        <Fragment key={uuid()}>
          <br />
          <Text
            typo={Typography.Size14}
          >
            { str }
          </Text>
        </Fragment>
      )
    })
  ), [])

  const titleComponent = useMemo(() => (
    <TitleWrapper className={contentClassName}>
      {
        leftIcon && (
          isIconName(leftIcon)
            ? (
              <IconWrapper
                color={leftIconColor}
                className={iconClassName}
                active={active}
                disableIconActive={disableIconActive}
              >
                <Icon
                  name={leftIcon}
                  size={IconSize.S}
                />
              </IconWrapper>
            ) : leftIcon
        )
      }
      <Title>
        {
          isString(content) ? (
            <Text
              typo={size === ListItemSize.XL
                ? Typography.Size18
                : Typography.Size14}
            >
              { content }
            </Text>
          ) : content
        }
      </Title>
    </TitleWrapper>
  ), [
    active,
    content,
    contentClassName,
    disableIconActive,
    iconClassName,
    leftIcon,
    leftIconColor,
    size,
  ])

  const descriptionComponent = useMemo(() => (
    <DescriptionWrapper
      active={active}
    >
      { leftIcon && <IconMargin /> }
      <Description descriptionMaxLines={descriptionMaxLines}>
        {
          isString(description) ? (
            <Text
              typo={Typography.Size14}
            >
              { getNewLineComponenet(description) }
            </Text>
          ) : description
        }
      </Description>
    </DescriptionWrapper>
  ), [
    active,
    description,
    descriptionMaxLines,
    getNewLineComponenet,
    leftIcon,
  ])

  const rightComponent = useMemo(() => (
    <RightContent>
      { rightContent }
    </RightContent>
  ), [rightContent])

  const ContentComponent = useMemo(() => (
    <>
      <ContentWrapper>
        { titleComponent }
        { description && descriptionComponent }
      </ContentWrapper>
      { rightContent && rightComponent }
    </>
  ), [
    titleComponent,
    description,
    descriptionComponent,
    rightContent,
    rightComponent,
  ])

  if (hide) return null

  if (!isNil(href)) {
    return (
      <Wrapper
        ref={forwardedRef}
        as="a"
        className={clazzName}
        size={size}
        draggable={false}
        href={href}
        target="_blank"
        rel="noopener noreferrer"
        onClick={handleClick}
        active={false}
        data-active={active}
        data-option-key={optionKey}
        data-testid={testId}
        {...othreProps}
      >
        { ContentComponent }
      </Wrapper>
    )
  }

  return (
    <Wrapper
      as={as}
      className={clazzName}
      size={size}
      onClick={handleClick}
      active={active}
      data-active={active}
      data-option-key={optionKey}
      data-testid={testId}
      {...othreProps}
    >
      { ContentComponent }
    </Wrapper>
  )
}

const ListItem = forwardRef(ListItemComponent)
ListItem.displayName = LIST_ITEM_COMPONENT_NAME

export default ListItem
