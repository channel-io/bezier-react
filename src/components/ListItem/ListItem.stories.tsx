/* External dependencies */
import React, {
  useState,
  useMemo,
} from 'react'
import base from 'paths.macro'

/* Internal dependencies */
import {
  iconList,
  getTitle,
} from '../../utils/storyUtils'
import ListItem from './ListItem'
import {
  ListItemSize,
  ListItemColorVariant,
} from './ListItem.types'

export default {
  title: getTitle(base),
  component: ListItem,
  argTypes: {
    width: {
      control: {
        type: 'range',
        min: 50,
        max: 500,
        step: 1,
      },
    },
    disabled: { control: { type: 'boolean' } },
    colorVariant: {
      control: {
        type: 'radio',
        options: [
          ...Object.values(ListItemColorVariant),
        ],
      },
    },
    active: { control: { type: 'boolean' } },
    leftIcon: { control: { type: 'select', options: [...iconList, undefined] } },
    size: { control: { type: 'select', options: ListItemSize } },
  },
}

const Template = ({ width, ...otherListItemProps }) => (
  <div style={{ width }}>
    <ListItem
      optionKey="menu-item-0"
      {...otherListItemProps}
    />
  </div>
)

export const Primary = Template.bind({})

Primary.args = {
  width: 388,
  size: ListItemSize.M,
  content: '상담이 열릴 때',
  description: '고객이 첫 메시지를 보내거나, 매니저가 상담을 다시 열거나, 자동으로 리오픈되면 트리거됩니다.',
  rightContent: '',
  leftIcon: 'inbox',
  active: false,
  focused: false,
  disableIconActive: false,
  descriptionMaxLines: 0,
  href: '',
}

export const MultiListItem = ({
  listRange,
}) => {
  const [activeIndex, setActiveIndex] = useState<Set<number>>(new Set())

  const isActive = (index: number) => activeIndex.has(index)

  const toggleActive = (index: number) => setActiveIndex((prevSet) => {
    if (prevSet.has(index)) {
      prevSet.delete(index)
      return new Set([...prevSet])
    }
    return new Set([...prevSet.add(index)])
  })

  const list = useMemo(() => [...Array(listRange).keys()], [listRange])

  return (
    <div>
      { list.map((index) => (
        <ListItem
          key={index}
          className={isActive(index) && 'active'}
          optionKey={`menu-item-${index}`}
          active={isActive(index)}
          onClick={() => toggleActive(index)}
          content={`이것은 ${index}번 아이템입니다.`}
        />
      )) }
    </div>
  )
}

MultiListItem.args = {
  listRange: 10,
}

MultiListItem.argTypes = {
  listRange: {
    control: {
      type: 'range',
      min: 2,
      max: 20,
    },
  },
}
