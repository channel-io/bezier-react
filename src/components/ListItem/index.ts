import ListItem from './ListItem'
import type ListItemProps from './ListItem.types'

export type {
  ListItemProps,
}

export {
  ListItem,
}
