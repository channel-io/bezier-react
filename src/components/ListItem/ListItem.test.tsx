/* External dependencies */
import React from 'react'
import { getWindow, getDocument } from 'ssr-window'

/* Internal dependencies */
import { render } from '../../utils/testUtils'
import ListItem, { LIST_ITEM_TEST_ID } from './ListItem'
import ListItemProps, { ListItemSize } from './ListItem.types'

describe('ListItem', () => {
  let props: ListItemProps

  beforeEach(() => {
    props = {
      content: 'this is content',
      optionKey: 'menu-item',
      active: false,
    }
  })

  const renderComponent = (optionProps?: Partial<ListItemProps>) => render(
    <ListItem {...props} {...optionProps} />,
  )

  it('should have "optionKey" value on "data-option-key" ', () => {
    const { getByTestId } = renderComponent({ optionKey: 'my-menu-item' })
    const rendered = getByTestId(LIST_ITEM_TEST_ID)

    expect(rendered).toHaveAttribute('data-option-key', 'my-menu-item')
  })

  it('should have "data-active" attribute when "active" prop is "true', () => {
    const { getByTestId } = renderComponent({ active: true })
    const rendered = getByTestId(LIST_ITEM_TEST_ID)

    expect(rendered).toHaveAttribute('data-active', 'true')
  })

  it('should have "a tag" related attributes  when "href" prop is string', () => {
    const { getByTestId } = renderComponent({ href: 'https://naver.com' })
    const rendered = getByTestId(LIST_ITEM_TEST_ID)
    expect(rendered).toHaveAttribute('href', 'https://naver.com')
    expect(rendered).toHaveAttribute('rel', 'noopener noreferrer')
    expect(rendered).toHaveAttribute('target', '_blank')
  })

  it('sholud not to be in the Document when "hide" prop is true', () => {
    const { queryByTestId } = renderComponent({ hide: true })
    const rendered = queryByTestId(LIST_ITEM_TEST_ID)
    expect(rendered).not.toBeInTheDocument()
  })

  describe('should have correct style of "size"', () => {
    it('size S', () => {
      const { getByTestId } = renderComponent({ size: ListItemSize.S })
      const rendered = getByTestId(LIST_ITEM_TEST_ID)
      expect(rendered).toHaveStyle('padding-top: 4px')
      expect(rendered).toHaveStyle('padding-bottom: 4px')
      expect(rendered).toHaveStyle('border-radius: 6px')
    })

    it('size M', () => {
      const { getByTestId } = renderComponent({ size: ListItemSize.M })
      const rendered = getByTestId(LIST_ITEM_TEST_ID)
      expect(rendered).toHaveStyle('padding-top: 6px')
      expect(rendered).toHaveStyle('padding-bottom: 6px')
      expect(rendered).toHaveStyle('border-radius: 6px')
    })

    it('size L', () => {
      const { getByTestId } = renderComponent({ size: ListItemSize.L })
      const rendered = getByTestId(LIST_ITEM_TEST_ID)
      expect(rendered).toHaveStyle('padding-top: 8px')
      expect(rendered).toHaveStyle('padding-bottom: 8px')
      expect(rendered).toHaveStyle('border-radius: 8px')
    })

    it('size XL', () => {
      const { getByTestId } = renderComponent({ size: ListItemSize.XL })
      const rendered = getByTestId(LIST_ITEM_TEST_ID)
      expect(rendered).toHaveStyle('padding-top: 10px')
      expect(rendered).toHaveStyle('padding-bottom: 10px')
      expect(rendered).toHaveStyle('border-radius: 12px')
    })
  })

  /* icon, text 등 contents와 padding에 따라 height 변합니다. */
  describe('should render correct height according to "size"', () => {
    it('size S', () => {
      renderComponent({ content: 'test', leftIcon: 'check', size: ListItemSize.S })
      const rendered = getDocument().querySelector(`div[data-testid=${LIST_ITEM_TEST_ID}]`)
      if (rendered) {
        const style = getWindow().getComputedStyle(rendered)
        setTimeout(() => expect(style.height).toBe('28px'), 0)
      }
    })

    it('size M', () => {
      renderComponent({ content: 'test', leftIcon: 'check', size: ListItemSize.M })
      const rendered = getDocument().querySelector(`div[data-testid=${LIST_ITEM_TEST_ID}]`)
      if (rendered) {
        const style = getWindow().getComputedStyle(rendered)
        setTimeout(() => expect(style.height).toBe('32px'), 0)
      }
    })

    it('size L', () => {
      renderComponent({ content: 'test', leftIcon: 'check', size: ListItemSize.L })
      const rendered = getDocument().querySelector(`div[data-testid=${LIST_ITEM_TEST_ID}]`)
      if (rendered) {
        const style = getWindow().getComputedStyle(rendered)
        setTimeout(() => expect(style.height).toBe('36px'), 0)
      }
    })

    it('size XL', () => {
      renderComponent({ content: 'test', leftIcon: 'check', size: ListItemSize.XL })
      const rendered = getDocument().querySelector(`div[data-testid=${LIST_ITEM_TEST_ID}]`)
      if (rendered) {
        const style = getWindow().getComputedStyle(rendered)
        setTimeout(() => expect(style.height).toBe('44px'), 0)
      }
    })
  })
})
