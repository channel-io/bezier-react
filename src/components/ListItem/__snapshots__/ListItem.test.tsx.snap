// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem Snapshot > 1`] = `
.c4 {
  font-size: 14px;
  line-height: 18px;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-property: color;
  transition-property: color;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  grid-row: 1;
  grid-column: 2;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 span {
  -webkit-transition-property: color;
  transition-property: color;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
}

.c3 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c1 {
  display: grid;
  grid-template-columns: fit-content(100%) minmax(0,1fr);
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 6px;
  overflow: hidden;
  border-radius: 6px;
  color: #000000D9;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  opacity: 1;
  -webkit-transition-property: background-color;
  transition-property: background-color;
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
}

.c0:hover {
  background-color: #0000000D;
}

.c0:hover .sc-iCoGMd {
  color: #00000066;
}

<div
  class="c0"
  data-active="false"
  data-option-key="menu-item"
  data-testid="bezier-react-list-menu-item"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <span
          class="c4"
          data-testid="bezier-react-text"
        >
          this is content
        </span>
      </div>
    </div>
  </div>
</div>
`;
